neovim-qt 0.2.16.1 (new formula)

gwerbin opened the pull request:
- [x] Have you followed the [guidelines for contributing](https://github.com/Homebrew/homebrew-core/blob/HEAD/CONTRIBUTING.md)?
- [x] Have you ensured that your commits follow the [commit style guide](https://docs.brew.sh/Formula-Cookbook#commit)?
- [x] Have you checked that there aren't other open [pull requests](https://github.com/Homebrew/homebrew-core/pulls) for the same formula update/change?
- [x] Have you built your formula locally with `brew install --build-from-source <formula>`, where `<formula>` is the name of the formula you're submitting?
- [ ] Is your test running fine `brew test <formula>`, where `<formula>` is the name of the formula you're submitting?
- [x] Does your build pass `brew audit --strict <formula>` (after doing `brew install --build-from-source <formula>`)? If this is a new formula, does it pass `brew audit --new <formula>`?

-----

Note: the application works fine (I've been using it on my own Tap for a while), but `brew test` fails, possibly due to some adverse interaction with the test sandbox:

```
% brew test neovim-qt
==> Testing neovim-qt
==> /usr/local/Cellar/neovim-qt/0.2.16.1/bin/nvim-qt --help
Killing child processes...
Error: neovim-qt: failed
Failure while executing; `/usr/bin/sandbox-exec -f /private/tmp/homebrew20210919-19842-p1ejzs.sb /usr/local/Homebrew/Library/Homebrew/vendor/portable-ruby/2.6.3_2/bin/ruby -W1 -- /usr/local/Homebrew/Library/Homebrew/test.rb /usr/local/Homebrew/Library/Taps/homebrew/homebrew-core/Formula/neovim-qt.rb` was terminated by uncaught signal TERM.
/usr/local/Homebrew/Library/Homebrew/sandbox.rb:142:in `exec'
/usr/local/Homebrew/Library/Homebrew/dev-cmd/test.rb:102:in `block (2 levels) in test'
/usr/local/Homebrew/Library/Homebrew/utils/fork.rb:45:in `block (3 levels) in safe_fork'
/usr/local/Homebrew/Library/Homebrew/utils/fork.rb:37:in `fork'
/usr/local/Homebrew/Library/Homebrew/utils/fork.rb:37:in `block (2 levels) in safe_fork'
/usr/local/Homebrew/Library/Homebrew/utils/fork.rb:34:in `open'
/usr/local/Homebrew/Library/Homebrew/utils/fork.rb:34:in `block in safe_fork'
/usr/local/Homebrew/Library/Homebrew/vendor/portable-ruby/2.6.3_2/lib/ruby/2.6.0/tmpdir.rb:93:in `mktmpdir'
/usr/local/Homebrew/Library/Homebrew/utils/fork.rb:33:in `safe_fork'
/usr/local/Homebrew/Library/Homebrew/dev-cmd/test.rb:90:in `block in test'
/usr/local/Homebrew/Library/Homebrew/dev-cmd/test.rb:45:in `each'
/usr/local/Homebrew/Library/Homebrew/dev-cmd/test.rb:45:in `test'
/usr/local/Homebrew/Library/Homebrew/brew.rb:129:in `<main>'
```

carlocab commented on Formula/neovim-qt.rb:
```diff
@@ -0,0 +1,32 @@
+      if OS.mac?
+        prefix.install "bin/nvim-qt.app"
+        bin.install_symlink prefix/"nvim-qt.app/Contents/MacOS/nvim-qt"
+      end
+
+      if OS.linux?
```
```suggestion
      else
```

carlocab commented on Formula/neovim-qt.rb:
```diff
@@ -0,0 +1,32 @@
+
+  def install
+    mkdir "build" do
+      system "cmake", "-DCMAKE_BUILD_TYPE=Release", ".."
```
```suggestion
      system "cmake", "..", *std_cmake_args
```

gwerbin replied:
Is `std_cmake_args` documented somewhere?

jgehrig replied:
Does `std_cmake_args` include `-DCMAKE_BUILD_TYPE=Release` (or RelWithDebugInfo)?

You'll definitely want to do a `Release` build of some kind...

An easy way to verify this is `nvim-qt --versoin`:
```
$ nvim-qt --version
NVIM-QT v0.2.17.9999
Build type: Release <-- Anything other than "Debug" is fine
...
```

carlocab commented on Formula/neovim-qt.rb:
```diff
@@ -0,0 +1,32 @@
+  end
+
+  test do
+    system bin/"nvim-qt", "--help"
```
Could we start an `nvim-qt` process, and then use `neovim-remote` to try to interact with it? That way we get to test a bit more than `--help`, which we strongly discourage for tests.

justinmk replied:
if it avoid some dependencies, you could also use just `nvim` instead of `neovim-remote` to interact

carlocab replied:
We could declare the dependency as test-only, so it should have no user-facing impact. Though if we do it directly with `nvim` that's fine too.

gwerbin replied:
Would using `neovim-remote` also prevent the test failure I reported in the OP?

carlocab replied:
It might; I'm not sure. We can try it out in CI to see.

gwerbin replied:
I realized that you don't need Neovim-remote to interact with Neovim. Recent versions of Neovim-Qt can pass arbitrary arguments along to the `nvim` program.

I copied the `test` block from the Neovim formula. It appears to work, but I still get the "Killing child processes"/"terminated by uncaught signal" issue. I wonder if this is a false error? Maybe Neovim-Qt sends SIGTERM to the underlying Neovim process?

Neovim-Qt has the ability to connect to an existing Neovim process with `nvim-qt --server`, which might avoid the SIGTERM problem. But then it doesn't support passing arguments to the underlying Neovim process!

carlocab commented:
Please run `brew style --fix neovim-qt` to fix the syntax errors, and then do `git rebase origin/master && git push --force` to get rid of the merge commits. (@BrewTestBot does not like merge commits, hence the `automerge-skip` label.)

gwerbin commented:
I fixed the style issues, but I'm still not sure how to test this effectively.

I added a test that uses Neovim-remote to interact with a Neovim instance that's started by Neovim-Qt. When I run the commands in a regular shell, everything seems to work perfectly. But when I run the test suite with `brew test`, somehow Neovim-remote is unable to connect to the socket established by the listening Neovim instance. I get the same problem if I use a filesystem socket or a hostname like `localhost:9999`:

```
% brew test --verbose Formula/neovim-qt.rb

==> Testing neovim-qt
/usr/local/Cellar/neovim-qt/0.2.16.1/bin/nvim-qt --nofork -- --listen /private/tmp/neovim-qt-test-20210921-6848-5827cy/nvim.sock

==> nvr --nostart --servername /private/tmp/neovim-qt-test-20210921-6848-5827cy/nvim.sock --remote /private/tmp/neovim-qt-test-20210921-6848-5827cy/test.txt

[!] Can't connect to: /private/tmp/neovim-qt-test-20210921-6848-5827cy/nvim.sock
```

carlocab commented:
Removed the lines that were too long so that CI can run properly.

carlocab commented:
The test seemed to run fine on ARM, but CI got cancelled again by another audit failure.

For future PRs, please do run `brew audit --strict <formula name>` (or `brew audit --new <formula name>`) before submitting, as that will save us all a bit of time. (Please _don't_ run `brew audit /path/to/formula.rb`.)

Is there no need to set `NVIM_QT_RUNTIME_PATH` when running `nvim-qt`, as documented, e.g., [here](https://github.com/equalsraf/neovim-qt/wiki/Build-Instructions#macos)?

carlocab commented:
I've pushed a fix for Linux, and decreased the amount of indentation needed in the `install` method for readability. This should be ready to merge soon when you've cleaned up the `test` block, plus one other thing (see review comment below).

carlocab commented on Formula/neovim-qt.rb:
```diff
@@ -0,0 +1,79 @@
+    system "cmake", "-S", ".", "-B", "build", *std_cmake_args, "-DUSE_SYSTEM_MSGPACK=ON"
+    system "cmake", "--build", "build"
+
+    if OS.mac?
+      prefix.install "build/bin/nvim-qt.app"
+      bin.install_symlink prefix/"nvim-qt.app/Contents/MacOS/nvim-qt"
+    else
+      bin.install "build/bin/nvim-qt"
+    end
```
We should check if there's actually an `install` target we can use, rather than just manually installing things ourselves.

gwerbin replied:
Yes, there is an `install` target. It is part of the Makefile that's generated by Cmake, and it uses `"-DCMAKE_INSTALL_PREFIX=#{prefix)"`, which appears to be [one of the `std_cmake_args`](https://rubydoc.brew.sh/Formula.html#std_cmake_args-instance_method).

Out of curiosity, why did you switch from `make` to `cmake --build build`? Is it OK if I reinstate the `mkdir "build" do` block so we can use the `make` and `make install`?

carlocab replied:
You can do the equivalent of `make install` with `cmake --install build`.

gwerbin replied:
@jgehrig sorry for my lack of CMake knowledge. What would the right `cmake --install` invocation be in this case? `cmake -S . --install`?

carlocab replied:
```
cmake --install build
```

gwerbin replied:
Thanks @carlocab . `cmake --install build` causes the Mac `.app` bundle to be installed to `bin/`, which causes an warning in `brew audit`. Is that OK?

It would be nice if we could install the bundle to `/Applications`. Maybe then we need a Cask instead of a Formula?

carlocab replied:
I see; we can just move things around after the fact. The same thing would be needed with `make install` anyway. Formulae cannot install outside of `HOMEBREW_PREFIX`, so we can't install anything into `/Applications`. We need a cask to do something like that.

gwerbin replied:
Noted! Maybe I can help the Nvim-Qt project set up a pipeline to automatically build and upload the app bundle.

carlocab replied:
Yes, that would be nice, and probably a much better experience for macOS users.

gwerbin commented:
> The test seemed to run fine on ARM, but CI got cancelled again by another audit failure.

Interesting, you don't get the SIGTERM error at all on that system? I might have to try and see if it works on Linux. I also just upgraded to Big Sur, so maybe it works there now 🤷 

> For future PRs, please do run `brew audit --strict <formula name>` (or `brew audit --new <formula name>`) before submitting, as that will save us all a bit of time. (Please _don't_ run `brew audit /path/to/formula.rb`.)

I've been doing `brew audit --strict --formula ./Formula/neovim-qt.rb`. Is that OK?

I didn't realize audit failures would actually block the CI, sorry about the wasted time.

> Is there no need to set `NVIM_QT_RUNTIME_PATH` when running `nvim-qt`, as documented, e.g., [here](https://github.com/equalsraf/neovim-qt/wiki/Build-Instructions#macos)?

I've never needed to do this. I think it's only for when you're running Neovim-Qt from a location other than its installation prefix. I asked on the issue tracker to confirm.

carlocab commented:
> Interesting, you don't get the SIGTERM error at all on that system? I might have to try and see if it works on Linux. I also just upgraded to Big Sur, so maybe it works there now 🤷

No `SIGTERM` on macOS at all. I think Linux is just from running things in a Docker container.

> I've been doing `brew audit --strict --formula ./Formula/neovim-qt.rb`. Is that OK?

Ideally not, as that will skip some checks (e.g. the ones that showed up in CI but you didn't see locally).

jgehrig commented:
> I think Linux is just from running things in a Docker container.

Yes, this seems like a reasonable explanation.

On Linux, `$DISPLAY` is required. For example, our CI has a workaround:
https://github.com/equalsraf/neovim-qt/blob/67cc4e414a8e64a475e55230818fab0f78415634/.github/workflows/build-test.yml#L141

Qt doesn't always play nice on headless Linux systems.

carlocab commented on Formula/neovim-qt.rb:
```diff
@@ -0,0 +1,79 @@
+    end
+  end
+
+  test do
```
```suggestion
  test do
    return if OS.linux? && ENV["HOMEBREW_GITHUB_ACTIONS"] # qt.qpa.xcb: could not connect to display
```
Let's just do this on Linux.

carlocab reviewed (approved):
Thanks, @gwerbin!
